rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/mod.rs

```diff
@@ -0,0 +1,131 @@
+    /// e.g. upon first starting up or if no sequenced data exists yet.
+    pub sequence: Option<u32>,
+    /// The block number at which the last sequence was indexed.
+    /// If the sequence is None, this can be thought of as the starting block
+    /// number to index from.
+    pub at_block: u32,
+}
+
+impl LastIndexedSnapshot {
+    fn next_target(&self) -> TargetSnapshot {
+        TargetSnapshot {
+            // If we haven't indexed anything yet, we start at 0, otherwise we increment.
+            sequence: self.sequence.map(|s| s + 1).unwrap_or(0),
+            at_block: self.at_block,
+        }
+    }
+
+    fn previous_target(&self) -> Option<TargetSnapshot> {
+        match &self.sequence {
+            // A previous target doesn't exist if we're trying to go backward
+            // from sequence 0 or if nothing has been indexed yet.
+            Some(0) | None => None,
+            Some(s) => Some(TargetSnapshot {
+                sequence: s.saturating_sub(1),
+                at_block: self.at_block,
+            }),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct TargetSnapshot {
+    pub sequence: u32,
+    pub at_block: u32,
+}
+
+#[derive(Debug)]
+pub enum SyncDirection {
+    Forward,
+    Backward,
+}
+
+/// A SequenceSyncCursor that syncs forwards in perpetuity.
+pub(crate) struct ForwardBackwardSequenceAwareSyncCursor<T> {
+    forward: ForwardSequenceAwareSyncCursor<T>,
+    backward: BackwardSequenceAwareSyncCursor<T>,
+    last_direction: SyncDirection,
+}
+
+impl<T: Sequenced + Debug> ForwardBackwardSequenceAwareSyncCursor<T> {
+    /// Construct a new contract sync helper.
+    pub async fn new(
+        latest_sequence_querier: Arc<dyn SequenceAwareIndexer<T>>,
+        db: Arc<dyn HyperlaneSequenceIndexerStore<T>>,
+        chunk_size: u32,
+        mode: IndexMode,
+    ) -> Result<Self> {
+        let (sequence_count, tip) = latest_sequence_querier
+            .latest_sequence_count_and_tip()
+            .await?;
+        let sequence_count = sequence_count.ok_or(ChainCommunicationError::from_other_str(
+            "Failed to query sequence",
+        ))?;
+        let forward_cursor = ForwardSequenceAwareSyncCursor::new(
+            chunk_size,
+            latest_sequence_querier.clone(),
+            db.clone(),
+            sequence_count,
+            tip,
+            mode,
+        );
+        let backward_cursor =
+            BackwardSequenceAwareSyncCursor::new(chunk_size, db, sequence_count, tip, mode);
+        Ok(Self {
+            forward: forward_cursor,
+            backward: backward_cursor,
+            last_direction: SyncDirection::Forward,
+        })
+    }
+}
+
+#[async_trait]
+impl<T: Sequenced + Debug> ContractSyncCursor<T> for ForwardBackwardSequenceAwareSyncCursor<T> {
+    async fn next_action(&mut self) -> Result<(CursorAction, Duration)> {
+        // TODO: Proper ETA for backwards sync
```

Comment:
These TODOs (& the ones in the Forward & Backward cursors) were present before & I'll leave that out of scope for now

Reply:
tracking this here https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3265